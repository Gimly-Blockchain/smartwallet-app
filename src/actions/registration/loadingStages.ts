import I18n from 'src/locales/i18n'
import en from '../../locales/en'

export const loadingStages = [
  I18n.t(en.ENCRYPTING_AND_STORING_DATA_LOCALLY),
  I18n.t(en.FUELING_WITH_ETHER),
  I18n.t(en.REGISTERING_DECENTRALIZED_IDENTITY),
  I18n.t(en.PREPARING_LAUNCH),
]
