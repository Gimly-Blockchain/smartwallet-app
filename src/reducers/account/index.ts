import { combineReducers } from 'redux'
import { did } from 'src/reducers/account/did'
import { claims } from 'src/reducers/account/claims'
import { loading } from 'src/reducers/account/loading'
import { SignedCredential } from 'jolocom-lib/js/credentials/signedCredential/signedCredential'
import { ClaimEntry } from 'jolocom-lib/js/credentials/credential/types'
import {
  CredentialOfferMetadata,
  CredentialOfferRenderInfo
} from 'jolocom-lib/js/interactionTokens/interactionTokens.types'

export interface DecoratedClaims {
  credentialType: string
  claimData: ClaimEntry
  id: string
  issuer: string
  subject: string
  expires?: Date,
  renderInfo?: CredentialOfferRenderInfo,
  metadata?: CredentialOfferMetadata,
  keyboardType?:
    | 'default'
    | 'number-pad'
    | 'decimal-pad'
    | 'numeric'
    | 'email-address'
    | 'phone-pad'
  renderInfo?: {
    // TODO: convert to CredentialRenderTypes
    renderAs?: string
    background?: {
      color?: string
      url?: string
    }
    logo?: {
      url: string
    }
    text?: {
      color: string
    }
  }
  // TODO: convert to CredentialOfferMetadata
  metadata?: {}
}

export interface CategorizedClaims {
  readonly [key: string]: DecoratedClaims[]
}

export interface ClaimsState {
  readonly loading: boolean
  readonly selected: DecoratedClaims
  readonly decoratedCredentials: CategorizedClaims
  readonly pendingExternal: SignedCredential[]
}

export interface DidState {
  readonly did: string
}

export interface LoadingState {
  readonly loading: boolean
}

// TODO avoid state.account.did.did access patterns
export interface AccountState {
  did: DidState
  claims: ClaimsState
  loading: LoadingState
}

export const accountReducer = combineReducers({
  did,
  claims,
  loading,
})
