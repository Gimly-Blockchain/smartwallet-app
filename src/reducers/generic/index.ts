import { AnyAction } from 'redux'
import { pickBy } from 'ramda'

export const APPWRAP_UPDATE_CONFIG = 'APPWRAP_UPDATE_CONFIG'
export const APPWRAP_SHOW_LOADER = 'APPWRAP_SHOW_LOADER'
export const APPWRAP_SET_LOCKED = 'APPWRAP_SET_LOCKED'
export const APPWRAP_REGISTER_CONFIG = 'APPWRAP_REGISTER_CONFIG'
export const APPWRAP_UNREGISTER_CONFIG = 'APPWRAP_UNREGISTER_CONFIG'

export interface AppWrapConfig {
  readonly withoutStatusBar: boolean
  readonly loading: boolean
  readonly dark: boolean
  readonly locked: boolean
}
const initialAppWrapAttrs: AppWrapConfig = {
  loading: false,
  locked: true,
  withoutStatusBar: false,
  dark: false,
}

export const isAppWrapConfigAttr = (key: string) => initialAppWrapAttrs[key] !== undefined
export const pickAppWrapConfigAttrs: (o: any) => AppWrapConfig =
  pickBy((v, k) => isAppWrapConfigAttr(k))

export interface AppWrapState {
  readonly appWrapConfig: AppWrapConfig
  readonly appWrapConfigsSet: AppWrapConfig[]
}
const initialAppWrapState: AppWrapState = {
  appWrapConfig: initialAppWrapAttrs,
  appWrapConfigsSet: []
}

// TODO(mnzaki): this code is unnecessarily complicated
// the complicated "ConfigsSet" keeping is to support nested <Wrapper>
// components
// - but we don't really need to support them
// - see 20APR[RN61.5] on notion.so:spaces/mnzaki

export const appWrapReducer = (
  state = initialAppWrapState,
  action: AnyAction,
): AppWrapState => {
  let appWrapConfig, appWrapConfigsSet
  switch (action.type) {
    case APPWRAP_REGISTER_CONFIG:
      const idx = state.appWrapConfigsSet.indexOf(action.value)
      if (idx > -1 || Object.keys(action.value).length === 0) return state
      appWrapConfig = {...state.appWrapConfig, ...action.value }
      appWrapConfigsSet = [...state.appWrapConfigsSet, action.value]
      return { appWrapConfig, appWrapConfigsSet }
    case APPWRAP_UNREGISTER_CONFIG:
      const idx2 = state.appWrapConfigsSet.indexOf(action.value)
      if (idx2 < 0) return state
      appWrapConfigsSet = state.appWrapConfigsSet.filter(s => s !== action.value)
      appWrapConfig = appWrapConfigsSet.reduce((prev, cur) => ({...prev, ...cur}), initialAppWrapAttrs)
      return { appWrapConfig, appWrapConfigsSet }

    case APPWRAP_SHOW_LOADER:
      return {...state, appWrapConfig: {...state.appWrapConfig, loading: action.value} }
    case APPWRAP_SET_LOCKED:
      return {...state, appWrapConfig: {...state.appWrapConfig, locked: action.value} }
    default:
      return state
  }
}
