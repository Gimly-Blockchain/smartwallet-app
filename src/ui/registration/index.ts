export { Loading } from './containers/loading'
export { Entropy } from './containers/entropy'
export { SeedPhrase } from './containers/seedPhrase'
