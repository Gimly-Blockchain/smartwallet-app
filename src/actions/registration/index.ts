import { navigationActions } from 'src/actions/'
import { routeList } from 'src/routeList'
import * as loading from 'src/actions/registration/loadingStages'
import { setDid } from 'src/actions/account'
import { JolocomLib } from 'jolocom-lib'
import { generateSecureRandomBytes } from 'src/lib/util'
import { ThunkAction } from '../../store'

const bip39 = require('bip39')

export const setLoadingMsg = (loadingMsg: string) => ({
  type: 'SET_LOADING_MSG',
  value: loadingMsg,
})

export const submitEntropy = (
  encodedEntropy: string,
): ThunkAction => dispatch => {
  dispatch(
    navigationActions.navigatorReset({
      routeName: routeList.Loading,
    }),
  )

  dispatch(setLoadingMsg(loading.loadingStages[0]))
  return dispatch(createIdentity(encodedEntropy))
}

export const startRegistration: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const randomPassword = await generateSecureRandomBytes(32)

  await backendMiddleware.keyChainLib.savePassword(
    randomPassword.toString('base64'),
  )

  return dispatch(
    navigationActions.navigate({
      routeName: routeList.Entropy,
    }),
  )
}

export const createIdentity = (encodedEntropy: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { encryptionLib, keyChainLib, storageLib, registry } = backendMiddleware

  const password = await keyChainLib.getPassword()
  const encEntropy = encryptionLib.encryptWithPass({
    data: encodedEntropy,
    pass: password,
  })
  const entropyData = { encryptedEntropy: encEntropy, timestamp: Date.now() }
  const userVault = JolocomLib.KeyProvider.fromSeed(
    Buffer.from(encodedEntropy, 'hex'),
    password,
  )

  dispatch(setLoadingMsg(loading.loadingStages[1]))

  await JolocomLib.util.fuelKeyWithEther(
    userVault.getPublicKey({
      encryptionPass: password,
      derivationPath: JolocomLib.KeyTypes.ethereumKey,
    }),
  )

  dispatch(setLoadingMsg(loading.loadingStages[2]))
  const identityWallet = await registry.create(userVault, password)

  const personaData = {
    did: identityWallet.identity.did,
    controllingKeyPath: JolocomLib.KeyTypes.jolocomIdentityKey,
  }

  dispatch(setDid(identityWallet.identity.did))
  dispatch(setLoadingMsg(loading.loadingStages[3]))
  await backendMiddleware.setIdentityWallet(userVault, password)

  await storageLib.store.encryptedSeed(entropyData)
  await storageLib.store.persona(personaData)

  return dispatch(
    navigationActions.navigatorReset({
      routeName: routeList.SeedPhrase,
      params: { mnemonic: bip39.entropyToMnemonic(encodedEntropy) },
    }),
  )
}
