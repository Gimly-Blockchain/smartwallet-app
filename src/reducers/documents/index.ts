import { AnyAction } from 'redux'
import { SET_EXPIRED_DOC, CLEAR_EXPIRED_DOC } from 'src/actions/documents'
import { DecoratedClaims } from '../account'

export interface DocumentsState {
  selectedExpiredDocument: DecoratedClaims
}

const initialState: DocumentsState = {
  selectedExpiredDocument: {
    credentialType: '',
    subject: '',
    id: '',
    issuer: '',
    expires: undefined,
    claimData: {
      type: '',
      documentNumber: '',
    },
  },
}

export const documentsReducer = (
  state = initialState,
  action: AnyAction,
): DocumentsState => {
  switch (action.type) {
    case SET_EXPIRED_DOC:
      return { ...state, selectedExpiredDocument: action.value }
    case CLEAR_EXPIRED_DOC:
      return {
        ...state,
        selectedExpiredDocument: initialState.selectedExpiredDocument,
      }
    default:
      return state
  }
}
