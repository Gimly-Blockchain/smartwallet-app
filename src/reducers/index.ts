import { combineReducers } from 'redux'
import { settingsReducer, SettingsState } from 'src/reducers/settings/'
import { accountReducer, AccountState } from 'src/reducers/account/'
import {
  registrationReducer,
  RegistrationState,
} from 'src/reducers/registration/'
import { navigationReducer } from 'src/reducers/navigation/'
import { NavigationState } from 'react-navigation'
import { ssoReducer, SsoState } from 'src/reducers/sso/'

export const rootReducer = combineReducers({
  settings: settingsReducer,
  account: accountReducer,
  registration: registrationReducer,
  navigation: navigationReducer,
  sso: ssoReducer,
})

export interface RootState {
  readonly settings: SettingsState
  readonly account: AccountState
  readonly registration: RegistrationState
  readonly navigation: NavigationState
  readonly sso: SsoState
}
