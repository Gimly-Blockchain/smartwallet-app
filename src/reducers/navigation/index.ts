import { Routes } from 'src/routes'

import {
  createNavigationReducer,
} from 'react-navigation-redux-helpers';


export const navigationReducer = createNavigationReducer(Routes)
