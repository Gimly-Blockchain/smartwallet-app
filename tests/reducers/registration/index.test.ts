import { registrationReducer } from 'src/reducers/registration/'

describe('registration reducer', () => {
  it('should initialize correctly', () => {
    expect(registrationReducer({}, { type: '@INIT' })).toMatchSnapshot()
  })
})
