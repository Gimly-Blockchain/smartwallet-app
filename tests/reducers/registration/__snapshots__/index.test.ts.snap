// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registration reducer should initialize correctly 1`] = `
Object {
  "loading": Object {
    "isRegistering": false,
    "loadingMsg": "Loading",
  },
}
`;
